### PR TITLE
Added a cloud renderer that uploads geometry to the GPU.

### PR DESCRIPTION
**Pull request by Zaggy1024**
#3578 updated to 1.12.x.

There are multiple commits on that previous pull request to see the changes over time to the cloud renderer.

### CODE REVIEW

**Comment by p455w0rd** [comment]
shouldn't en_US.lang be en_us.lang in 1.12? or does that only apply to mods?

**Comment by liach** [comment]
@p455w0rd It seems this is an existing bug in forge.

**Comment by Zaggy1024** [comment]
Don't look at me, all I did was edit the file. :P

**Comment by mezz** [comment]
Tested.

Before:


After:


I noticed the vanilla clouds and fastclouds do not line up when you switch modes, is that intentional or is there an issue?

**Comment by mezz** [comment]
I also see this in the log, it should be resolved somehow:
```
[21:34:31] [main/ERROR] [FML]: Unable to determine registrant mod for net.minecraftforge.client.CloudRenderer@78029cf3. This is a critical error and should be impossible
java.lang.Throwable: null
	at net.minecraftforge.fml.common.eventhandler.EventBus.register(EventBus.java:74) [EventBus.class:?]
	at net.minecraftforge.client.CloudRenderer.<init>(CloudRenderer.java:84) [CloudRenderer.class:?]
	at net.minecraftforge.fml.client.FMLClientHandler.onInitializationComplete(FMLClientHandler.java:463) [FMLClientHandler.class:?]
	at net.minecraft.client.Minecraft.init(Minecraft.java:552) [Minecraft.class:?]
	at net.minecraft.client.Minecraft.run(Minecraft.java:377) [Minecraft.class:?]
	at net.minecraft.client.main.Main.main(Main.java:118) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
	at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97) [Forge_main/:?]
	at GradleStart.main(GradleStart.java:26) [Forge_main/:?]
```

**Comment by mezz** [comment]
I also see this in the log, it should be resolved somehow:
```
[21:34:31] [main/ERROR] [FML]: Unable to determine registrant mod for net.minecraftforge.client.CloudRenderer@78029cf3. This is a critical error and should be impossible
java.lang.Throwable: null
	at net.minecraftforge.fml.common.eventhandler.EventBus.register(EventBus.java:74) [EventBus.class:?]
	at net.minecraftforge.client.CloudRenderer.<init>(CloudRenderer.java:84) [CloudRenderer.class:?]
	at net.minecraftforge.fml.client.FMLClientHandler.onInitializationComplete(FMLClientHandler.java:463) [FMLClientHandler.class:?]
	at net.minecraft.client.Minecraft.init(Minecraft.java:552) [Minecraft.class:?]
	at net.minecraft.client.Minecraft.run(Minecraft.java:377) [Minecraft.class:?]
	at net.minecraft.client.main.Main.main(Main.java:118) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
	at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97) [Forge_main/:?]
	at GradleStart.main(GradleStart.java:26) [Forge_main/:?]
```

**Comment by mezz** [comment]
@Zaggy1024 any update?

**Comment by Zaggy1024** [comment]
I'm not seeing either of those issues in my development environment... The clouds appear to be identical if I make it switch them off and on every half second.

The only errors I'm seeing are for model loading that's unrelated to the cloud renderer.

File: src/main/java/net/minecraftforge/client/CloudRenderer.java [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+
+    private int texW;
+    private int texH;
+
+    public CloudRenderer()
+    {
+        // Resource manager should always be reloadable.
+        ((IReloadableResourceManager) mc.getResourceManager()).registerReloadListener(this);
+
+        MinecraftForge.EVENT_BUS.register(this);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/client/CloudRenderer.java]: I tried again with a new workspace but this line still causes an error log.
There is no active mod container when this is called.

```
[main/ERROR] [FML]: Unable to determine registrant mod for net.minecraftforge.client.CloudRenderer@72f05bd9. This is a critical error and should be impossible
java.lang.Throwable: null
	at net.minecraftforge.fml.common.eventhandler.EventBus.register(EventBus.java:74) [EventBus.class:?]
	at net.minecraftforge.client.CloudRenderer.<init>(CloudRenderer.java:84) [CloudRenderer.class:?]
	at net.minecraftforge.fml.client.FMLClientHandler.onInitializationComplete(FMLClientHandler.java:463) [FMLClientHandler.class:?]
	at net.minecraft.client.Minecraft.init(Minecraft.java:552) [Minecraft.class:?]
	at net.minecraft.client.Minecraft.run(Minecraft.java:377) [Minecraft.class:?]
	at net.minecraft.client.main.Main.main(Main.java:118) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.12.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.12.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_131]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_131]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_131]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_131]
	at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97) [Forge_main/:?]
	at GradleStart.main(GradleStart.java:26) [Forge_main/:?]
```

I think it has to be called from somewhere when FML is the active mod container.
Make sure you are testing with your branch merged into the latest for the 1.12.x branch.

**Review by mezz — CHANGES_REQUESTED** [review]
Tested again with a fresh workspace and the clouds line up fine but there is still a console error.

**Comment by bs2609** [comment]
Any update on this?

**Comment by Zaggy1024** [comment]
I was never able to reproduce the issue... I'm going to try rebasing one more time and see if it happens this time, but if not, I don't know what I can do about it, except trial and error having someone else test it.

**Comment by Zaggy1024** [comment]
Actually, it looks like what's happening on your end shouldn't even be possible, since it's stepping through the stack checking for packages owned by a mod, and it's not made to ignore the class that's searching for the package owners (in other words, it finds itself), so it should *always* find the owner as FML, if fml.common is part of the FML mod container. I'm not sure how this code was meant to work, but it seems to me it needs to be made to ignore a certain number of classes off the top of the stack trace (varying based on where it's being called from in FML code)...

Either way, I'm not sure why it would fail on your end, it might be worth debugging in your IDE to see what it's doing. The relevant code is https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/fml/common/LoadController.java#L388.

In the meantime, I'll see if I can find a place where I can call the constructor from. I believe there was a reason I picked that method rather than the more standard ones in FMLClientHandler, but I can't remember what that reason was.

**Comment by Zaggy1024** [comment]
Actually, I'm wondering if it would be best to register it in FMLContainer. Is that bad practice or would that work? Another option I guess would be to set the active mod container in FMLClientHandler, but I'm not sure of the best way to get the instance of FMLContainer to set it to.

**Comment by LexManos** [comment]
ForgeInternalHandler

File: src/main/java/net/minecraftforge/client/CloudRenderer.java [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+        }
+    }
+
+    private void build()
+    {
+        Tessellator tess = Tessellator.getInstance();
+        BufferBuilder buffer = tess.getBuffer();
+
+        if (OpenGlHelper.useVbo())
+            vbo = new net.minecraft.client.renderer.vertex.VertexBuffer(FORMAT);
```
Comment (bs2609) [inline src/main/java/net/minecraftforge/client/CloudRenderer.java]: You don't need the fully-qualified name here now that this is targeting 1.12 - might as well take advantage of the name change.

**Comment by bs2609** [comment]
@mezz, mind looking at this again now it's been updated?

**Comment by mezz** [comment]
Tested and looks good to me.

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
🎉

**Pull request merged**